### PR TITLE
Skip fee policies for in-market orders

### PR DESCRIPTION
**Pull request by sunce86**
# Description
A subset of changes from https://github.com/cowprotocol/services/pull/2188 (part 2)

# Changes

- [ ] Move fee policies to `domain`
- [ ] No fee policies for in-market limit orders

## How to test
Existing tests.

### CODE REVIEW

File: crates/autopilot/src/infra/persistence/api/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/api/mod.rs:4]: What is the reasoning behind the api submodule and this structuring? (does it contain infra code that is used by the API, which API specifically)?

I'm not a fan of structuring the modules this way. Not having a single domain facing persistence modules makes it very hard to see what persistence the autopilot requires (I always struggle navigating our db queries because they are scattered across many different files). 
If the top level `Persistence` module becomes to big, I think we should turn it into a lightweight facade that dispatches the work into submodules that are organized by the type of persistence that is being used (e.g. persistence::s3 vs. persistence::postgres). If then the postgres module is also becoming too big, maybe we can think about extensions that are named after the type of queries they perform.

File: crates/autopilot/src/infra/persistence/dto/order.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/dto/order.rs:38]: Is this change related?

**Review by fleupold — COMMENTED** [review]
Let's keep the boundary logic as small as possible (I think a lot of the logic should live in the solvable_orders.rs module)

File: crates/autopilot/src/infra/persistence/dto/quote.rs
Comment (MartinquaXD) [inline crates/autopilot/src/infra/persistence/dto/quote.rs:7]: I'm a bit surprised to see `dto` in the path here.
My understanding is that `dto`s are only dumb data objects that can be (de)serialized to send them to other services via HTTP requests or some other means. In this case it seems more appropriate to me to have these functions as part of the module that implements the operations on the DB.

File: crates/autopilot/src/infra/persistence/dto/order.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/dto/order.rs:38]: Yes, [I erased the `Order` type from `driver_model`](https://github.com/cowprotocol/services/pull/2250/files#diff-53c95c46cb3be0831e6245ae68a96c00e93973b8308f6036f4562de00b3b27c1L127-L131) and reused this `persistence::dto::Order` which is using `OrderClass` which needs to be flattened to make it equal.

File: crates/autopilot/src/boundary/auction.rs
Comment (sunce86) [reply crates/autopilot/src/boundary/auction.rs]: Moved auction conversion to solvable orders.

File: crates/autopilot/src/boundary/order.rs
Comment (sunce86) [reply crates/autopilot/src/boundary/order.rs]: Moved.

File: crates/autopilot/src/database/auction.rs
Comment (sunce86) [reply crates/autopilot/src/database/auction.rs]: Reverted, agreed.

File: crates/autopilot/src/database/auction.rs
Comment (sunce86) [reply crates/autopilot/src/database/auction.rs]: Simplified to `HashMap<domain::OrderUid, domain::Quote>`.

File: crates/autopilot/src/domain/fee/policy.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs]: not anymore, good catch

File: crates/autopilot/src/domain/fee/policy.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs:14]: renamed

File: crates/autopilot/src/infra/persistence/dto/quote.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/dto/quote.rs:21]: renamed.

File: crates/autopilot/src/domain/fee/policy.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs]: Yes, added

File: crates/autopilot/src/infra/persistence/dto/quote.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/dto/quote.rs]: `boundary::DatabaseQuote` is kind of a `dto` object between autopilot and database, since it's a serializable/deserializable representation of the `domain::Quote`.

I wanted to have this way of structuring paths so I can refer to `to_domain/from_domain` methods as `dto::object::to_domain()` for example, which reads much more nicely than `object::dto::to_domain()` at least IMO

File: crates/autopilot/src/infra/persistence/api/mod.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/api/mod.rs]: refactored.

Comment (fleupold) [reply crates/autopilot/src/infra/persistence/api/mod.rs]: I see that this will likely become too big if we move all db queries. In this case, I'd still have one domain facing struct `Persistence` which just dispatches the calls to sub-persistence solutions (postgres, s3, etc) and use logically sepearated modules within each persistence type.

For now I think the refactored version is ok.

File: crates/autopilot/src/domain/fee/policy.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs]: Does that mean that this function also does conversion from `boundary::Order` to `domain::Order`?

File: crates/autopilot/src/domain/quote/mod.rs
Comment (fleupold) [inline crates/autopilot/src/domain/quote/mod.rs:14]: Are all these fields relevant for the autopilot domain? If not, can we reduce to the ones that matter?

File: crates/autopilot/src/infra/persistence/dto/order.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/dto/order.rs:271]: What do you think about using a model here that can actually be written straight into the db (which I believe this enum cannot)?

File: crates/autopilot/src/domain/fee/policy.rs
Comment (fleupold) [reply crates/autopilot/src/domain/fee/policy.rs]: I think so, although really we should always have a plan for how to phase out boundary (if I understand correctly this  module is supposed to be temporary and a tool to keep the scope of the refactoring small).

We load raw order data from the database (without fee policies). These need to be represented in the domain somehow (e.g. Orders) and shouldn't stay `boundary::Order` forever. Then it is IMO part of the domain logic to add fees to this type, which likely should yield a new domain type (PreparedOrders or something like this), which are then type used on `domain::Auction`.

Turning a Oder into a PreparedOrder is to me what the domain::Fee component should do (until we have other prep work that needs to be done by the autopilot).

One could potentially also take `domain::Order` instance with fee_policy being `[]` and mutate them here, but this could easily lead to us believing we have a "prepared auction" whereas we missed an important step of the preparation (adding fees)

File: crates/autopilot/src/domain/quote/mod.rs
Comment (sunce86) [reply crates/autopilot/src/domain/quote/mod.rs]: You are right. reduced.

File: crates/autopilot/src/infra/persistence/dto/order.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/dto/order.rs:271]: Should be shared probably, so I will link this to be tackled in https://github.com/cowprotocol/services/pull/2118 once this PR is merged.

File: crates/autopilot/src/database/quotes.rs
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: reused.

File: crates/autopilot/src/domain/fee/policy.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/policy.rs]: To me, we should introduce new state of the order (in your case `PreparedOrder`) only if, for example, somewhere in the middle of the autopilot domain order lifecycle, the state of the order changes (think state machine). For now, I see this as potentially needed only for `order.executed` field, but not even for this, since the state is saved to database after each trade.

IMO, what you want should be part of refactoring the solvable orders. So, there, we should have a different implementation of fetching the data from the DB (instead of OPEN_ORDERS) and with that, we should define a new dto (as a replacement for `model::Order`) and use this dto to convert to domain::Order (and that conversion would live in `infra`)

File: crates/autopilot/src/boundary/auction.rs
Comment (sunce86) [reply crates/autopilot/src/boundary/auction.rs]: AFAIS we already use it this way across codebase.

File: crates/autopilot/src/database/auction.rs
Comment (sunce86) [reply crates/autopilot/src/database/auction.rs]: I think we already mentioned this somewhere. Anyway, implementing it in a single query right now is double work, since we also want to refactor solvable orders so that it:
1. fetch open orders but just a subset of data (compared to current OPEN ORDERS) needed for autopilot.
2. fetch other needed data in the same single query (quotes, whatever else).

So my plan was to merge it as is, and observe the performance and if ok, leave it that way until refactoring of solvable orders.

File: crates/autopilot/src/database/quotes.rs
Comment (sunce86) [reply crates/autopilot/src/database/quotes.rs]: Copy of ☝️

**Comment by sunce86** [comment]
I think I addressed most of the comments, so requesting review again.

File: crates/autopilot/src/domain/fee/policy.rs
Comment (fleupold) [reply crates/autopilot/src/domain/fee/policy.rs]: > we should define a new dto (as a replacement for model::Order) and use this dto to convert to domain::Order (and that conversion would live in infra)

Would `domain::Order` in this world be without the fee? Applying fee logic to me is a core part of the autopilot domain and should thus not live in infra. The reason why I wouldn't want to reuse the same model (say db returns `Order` with `fee: vec![]` and then the fee domain component amends it), is that it's not clear from the type if an object is "ready" to be passed along to e.g. the `request_solve` method.

File: crates/autopilot/src/database/auction.rs
Comment (fleupold) [reply crates/autopilot/src/database/auction.rs]: Isn't it trivial though to expose a `read_quotes` method that fetches everything in a single query so we avoid up to 2k roundtrips per order? You mentioned sharing the connection somewhere before as a performance improvement (which made a lot of sense to me), but if we still do 2k roundtrips reusing the connection isn't going to help much, is it?

With observing performance the issue is that we will only see problems in prod, so we may loose valuable time to shipping.

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/mod.rs:52]: nit: this method can directly take dto::auction, no need to convert twice.

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (fleupold) [inline crates/autopilot/src/infra/persistence/mod.rs:105]: is this error type used?

**Review by fleupold — APPROVED** [review]
Looks good to me, although I do think we should walk the extra mile and expose a `read_orders` (plural) method on the db. I don't feel like we should take the risk of running into performance issues in prod when we want to launch.

File: crates/autopilot/src/domain/fee/mod.rs
Comment (MartinquaXD) [inline crates/autopilot/src/domain/fee/mod.rs:45]: nit: This gets  rid of the unwraps further down.
```suggestion
                let Some(quote) = quote else {
```

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (MartinquaXD) [inline crates/autopilot/src/infra/persistence/mod.rs:51]: nit: This would remove 1 level of indentation.
```suggestion
        let Some(uploader) = self.s3.clone() else {
            return;
        };
```

File: crates/autopilot/src/solvable_orders.rs
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:80]: nit: only holds a single policy at the moment. Same applies to other variable names in this file.
```suggestion
    fee_policy: domain::ProtocolFee,
```

File: crates/autopilot/src/solvable_orders.rs
Comment (MartinquaXD) [inline crates/autopilot/src/solvable_orders.rs:237]: nit: Doesn't make a huge difference in this case because there is not a lot of contention on this lock but I think it's a good habit to prepare the data you want to store using the Mutex before you take the lock to keep the critical section as short as possible.

**Review by MartinquaXD — APPROVED** [review]
Just nits remaining.

File: crates/autopilot/src/domain/fee/mod.rs
Comment (sunce86) [reply crates/autopilot/src/domain/fee/mod.rs]: Nice

File: crates/autopilot/src/solvable_orders.rs
Comment (sunce86) [reply crates/autopilot/src/solvable_orders.rs]: fixed

File: crates/autopilot/src/solvable_orders.rs
Comment (sunce86) [reply crates/autopilot/src/solvable_orders.rs]: Makes sense, extracted

File: crates/autopilot/src/infra/persistence/mod.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/mod.rs]: not anymore, good catch.

File: crates/autopilot/src/database/auction.rs
Comment (sunce86) [reply crates/autopilot/src/database/auction.rs]: If we want a single query will need to build the query manually. To unblock https://github.com/cowprotocol/services/pull/2118 I am starting to work on this in a separate PR.

File: crates/autopilot/src/domain/auction/order.rs
Comment (sunce86) [reply crates/autopilot/src/domain/auction/order.rs:30]: https://github.com/cowprotocol/services/issues/2256

File: crates/autopilot/src/database/auction.rs
Comment (sunce86) [reply crates/autopilot/src/database/auction.rs]: https://github.com/cowprotocol/services/issues/2257

File: crates/autopilot/src/infra/persistence/dto/order.rs
Comment (sunce86) [reply crates/autopilot/src/infra/persistence/dto/order.rs:271]: mentioned https://github.com/cowprotocol/services/pull/2118#issuecomment-1883224848

**Pull request merged**
